File: competitive-programming/src/edu/algos/li17_graph/medium/_MultiBfs_L286_WallsAndGates.java
package edu.algos.li17_graph.medium;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class _MultiBfs_L286_WallsAndGates {
    static int INF = Integer.MAX_VALUE;

    public static void main(String[] args) {
        _MultiBfs_L286_WallsAndGates obj = new _MultiBfs_L286_WallsAndGates();
        int[][] rooms = new int[][] {
                { INF, -1, 0, INF },
                { INF, INF, INF, -1 },
                { INF, -1, INF, -1 },
                { 0, -1, INF, INF } };
        obj.wallsAndGates(rooms);

        for (int[] row : rooms) {
            System.out.println(Arrays.toString(row));
        }
    }

    // Problem: https://leetcode.com/problems/walls-and-gates/
    // https://www.lintcode.com/problem/663/
    // Idea: Do multi level bfs from all the gates and track dist variable
    // Increase the dist for each level.
    public void wallsAndGates(int[][] rooms) {
        int m = rooms.length, n = rooms[0].length;

        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (rooms[i][j] == 0) {
                    q.add(new int[] { i, j });
                    visited[i][j] = true;
                }
            }
        }
        int dist = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            for (int k = 0; k < size; k++) {
                int[] room = q.remove();
                int row = room[0], col = room[1];
                rooms[row][col] = dist;
                int[][] directions = new int[][] { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };
                for (int[] dr : directions) {
                    int nrow = row + dr[0];
                    int ncol = col + dr[1];
                    if (nrow >= 0 && nrow < m && ncol >= 0 && ncol < n 
                            && rooms[nrow][ncol] != -1 && !visited[nrow][ncol]) {
                        visited[nrow][ncol] = true;
                        q.add(new int[] { nrow, ncol });
                    }

                }
            }
            dist++;
        }
    }

}

File: dsa/patterns/13_graphs/python/multi_bfs_l286_walls_and_gates.py
from collections import deque

INF = 2**31 - 1


def walls_and_gates(rooms):
    """
    Problem:
    https://leetcode.com/problems/walls-and-gates/
    https://www.lintcode.com/problem/663/

    Idea:
    - This is a classic **multi-source BFS** problem.
    - All gates (cells with value 0) act as BFS sources.
    - Start BFS simultaneously from all gates.
    - Each BFS level increases distance by 1.
    - Update empty rooms (INF) with the minimum distance to a gate.
    - Walls (-1) are never visited.

    Time Complexity: O(m * n)
    Space Complexity: O(m * n)
    """
    if not rooms or not rooms[0]:
        return

    m, n = len(rooms), len(rooms[0])
    q = deque()

    # Initialize queue with all gates
    for i in range(m):
        for j in range(n):
            if rooms[i][j] == 0:
                q.append((i, j, 0))  # (row, col, distance)

    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    while q:
        row, col, dist = q.popleft()

        for dr, dc in directions:
            nr, nc = row + dr, col + dc
            if 0 <= nr < m and 0 <= nc < n and rooms[nr][nc] == INF:
                rooms[nr][nc] = dist + 1
                q.append((nr, nc, dist + 1))


if __name__ == "__main__":
    rooms = [
        [INF, -1, 0, INF],
        [INF, INF, INF, -1],
        [INF, -1, INF, -1],
        [0, -1, INF, INF],
    ]

    walls_and_gates(rooms)

    for row in rooms:
        print(row)
